test/com/esotericsoftware/kryo/serializers/SerializerPriorityTest.java

```diff
@@ -0,0 +1,126 @@
+
+package com.esotericsoftware.kryo.serializers;
+
+import com.esotericsoftware.kryo.DefaultSerializer;
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.KryoSerializable;
+import com.esotericsoftware.kryo.Serializer;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.esotericsoftware.kryo.serializers.DefaultSerializers.KryoSerializableSerializer;
+import com.esotericsoftware.kryo.serializers.DefaultSerializers.TreeMapSerializer;
+import org.junit.Test;
+
+import java.util.TreeMap;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * @author Niels Basjes
+ */
+@SuppressWarnings("synthetic-access")
+public class SerializerPriorityTest {
+
+    // Dummy serializer
+    public static final class MyAnnotationSerializer extends Serializer<Object> {
+        @Override
+        public void write(Kryo kryo, Output output, Object object) {
+        }
+
+        @Override
+        public Object read(Kryo kryo, Input input, Class type) {
+            return null;
+        }
+    }
+
+    // Dummy KryoSerializable implementation
+    public interface BaseKryoSerializable extends KryoSerializable {
+        @Override
+        default void write(Kryo kryo, Output output) {
+        }
+
+        @Override
+        default void read(Kryo kryo, Input input) {
+        }
+    }
+
+    public static final class MyTreeMap extends TreeMap<String, String> {   }
+    public static final class MyKryoSerializable implements BaseKryoSerializable {    }
+    public static final class MyTreeMapKryoSerializable extends TreeMap<String, String> implements BaseKryoSerializable {    }
+
+    @DefaultSerializer(MyAnnotationSerializer.class)
+    public static final class MyAnnotated {    }
+
+    @DefaultSerializer(MyAnnotationSerializer.class)
+    public static final class MyAnnotatedTreeMap extends TreeMap<String, String> {    }
+
+    @DefaultSerializer(MyAnnotationSerializer.class)
+    public static final class MyAnnotatedKryoSerializable implements BaseKryoSerializable {    }
+
+    @DefaultSerializer(MyAnnotationSerializer.class)
+    public static final class MyAnnotatedTreeMapKryoSerializable extends TreeMap<String, String> implements BaseKryoSerializable {    }
+
+    private void checkCorrectSerializer(Class<?> myClass, Class<? extends Serializer<?>> expectedSerializer) {
+        Kryo kryo = new Kryo();
+        kryo.register(myClass);
+        assertEquals(expectedSerializer, kryo.getSerializer(myClass).getClass());
+    }
+
+    @Test
+    public void testTreeMap() {
+        checkCorrectSerializer(MyTreeMap.class, TreeMapSerializer.class);
+    }
+
+    @Test
+    public void testKryoSerializable() {
+        checkCorrectSerializer(MyKryoSerializable.class, KryoSerializableSerializer.class);
+    }
+
+    @Test
+    public void testTreeMapKryoSerializable() {
+        // We choose that the KryoSerializable is more important than the default Serializer
```

Comment:
Instead of comments you could also adjust the test names. Something like `testKryoSerializableSerializerHasHigherPriorityThanDefaultSerializer`. This doesn't look great but its much easier to see what's going on when the test fails.

Reply:
Done.

Reply:
Please remove the comments now that the test names reflect the test purpose.

Apart from that LGTM.

Reply:
Done